### PR TITLE
FIX-#3619: Fix ValueError when doing loc on empty dataframe

### LINKED ISSUE #3619: "ValueError: Invalid integer data type 'f'." when doing `loc` on empty dataframe
We currently default to pandas when we're operating on empty dataframes, but this doesn't happen for some properties. `loc` will fail if you try to use it on an empty dataframe.

Reproduction script:

```python
import modin.pandas as pd
import ray
import pandas_market_calendars as mcal
from pandas.tseries.offsets import BusinessDay

import datetime as dt
ENTITIES = ['037833100', '594918104', '023135106', '30231G102', '478160104', '30303M102', '369604103', '00206R102',
          '46625H100', '742718109', '02079K305', '949746101', '931142103', '92343V104', '717081103', '084670702',
          '166764100', '191216100', '458140100', '58933Y105', '68389X105', '060505104', '20030N101', '437076102',
          '92826C839', '17275R102', '718172109', '713448108', '254687106', '459200101', '172967424', '91324P102',
          '031162100', '02209S103', '57636Q104', '88579Y101', '500754106', '806857108', '375558103', '00287Y109',
          '580135101', '126650100', '747525103', '110122108', '931427108', '532457108', '438516106', '913017109',
          '151020104', '097023105', '855244109', '907818108', '654106103', '911312106', '902973304', '741503403',
          '16119P108', '539830109', '882508104', '761713106', 'Y09827109', '38141G104', '609207105', '09062X103',
          '22160K105', '828806109', '194162103', '548661107', '026874784', '002824100', '617446448', '09247X101',
          '260543103', '025816109', '883556102', '887317303', '263534109', '65339F101', '293792107', '26441C204',
          '674599105', '235851102', '842587107', '00724F101', '26875P101', '37045V100', '79466L302', '872540109',
            '49456B101', '20825C104', '59156R108', '70450Y103', '149123101', '25746U109', '345370860', '03027X100',
            '369550108', '517834107', '61166W101', '693475105']


def main():
    nyse = mcal.get_calendar('NYSE')

    trade_date_range = pd.bdate_range(
        end=dt.date.fromisoformat("2017-01-10") - BusinessDay(n=1),
        periods=10,
        freq="C",
        holidays=list(nyse.holidays().holidays),
    )
    trade_date_range.name = "pricing_date"

    pricing_index = pd.MultiIndex.from_product([trade_date_range, ENTITIES], names=["pricing_date", "entity"])

    df = pd.DataFrame(index=pricing_index)

    print(df.loc['2017-01-09'].index.get_level_values("entity"))


if __name__ == '__main__':
    ray.init()
    main()
```

### PR DESCRIPTION
**Pull request by naren-ponder**
Signed-off-by: Naren Krishna <naren@ponder.io>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3619 
- [x] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

**Review by devin-petersohn — COMMENTED** [review]
Maybe we can do something like:

```python
if self.df.empty:
    return self.df._default_to_pandas(lambda df: df.loc[key])
```

It's similar to your implementation, but will work on row label objects too I believe

**Review by devin-petersohn — COMMENTED** [review]
This looks great, a couple of comments:

1.) Can we do this for `iloc` as well?
2.) Let's add a test to make sure we have the correct behavior

File: modin/pandas/indexing.py [hunk 1]
```diff
@@ -528,6 +528,8 @@ def __getitem__(self, key):
             else:
                 result_slice = self.df.columns.slice_locs(col_loc.start, col_loc.stop)
                 return self.df.iloc[:, slice(*result_slice)]
+        if self.df.empty:
+            return self.df._default_to_pandas(lambda df: df.loc[key])
```
Comment (dchigarev) [inline modin/pandas/indexing.py:532]: maybe we should put this at the top of the function, so we make sure that the execution won't go on that probably-failing fast path above?

Comment (naren-ponder) [reply modin/pandas/indexing.py]: Yes, good idea.

**Comment by dchigarev** [comment]
do the `loc/iloc` assignments perform correctly on empty frames, or we should default to pandas on `__setitem__` as well?

**Comment by naren-ponder** [comment]
@dchigarev Can you re-run the TeamCity Dask test? I believe it's a flake.

**Comment by dchigarev** [comment]
> @dchigarev Can you re-run the TeamCity Dask test? I believe it's a flake.

done, now it's a green

**Comment by naren-ponder** [comment]
@dchigarev Added tests for `loc` and `iloc` assignments on empty dataframes. Those seem to work without any code changes on my end. @devin-petersohn `iloc` actually seems to work for the get and set without any code changes.

**Comment by dchigarev** [comment]
> iloc actually seems to work for the get and set without any code changes, so the only change necessary is for the loc get method.

BTW, the actual problem of why the reported bug occurs is the following way of computing a column lookup for `.loc`, for empty columns and `slice(None)` indexer it returns an empty list:
https://github.com/modin-project/modin/blob/41213581b974760928ed36f05c479937dcdfea55/modin/pandas/indexing.py#L698-L701
Pandas backend is failing to process `[]` instead of a `slice(None)` since for non-full-axis grabs it tries to do some partitioning stuff under an empty frame and fails.

It's actually pretty lucky, that the scenario from reproducer works well for the `.iloc` because the only difference from the `.loc` flow is the `_compute_lookup`, that suddenly returns a `slice(None)` indexer for an empty frame and full-axis grab, which is acceptable for <i>pandas</i> backend since it doesn't touch partitioning on a `slice(None)` indexer:
https://github.com/modin-project/modin/blob/41213581b974760928ed36f05c479937dcdfea55/modin/pandas/indexing.py#L817-L826

So, actually, the reason why `iloc.__getitem__` on an empty frame is working is based on two inconsistencies:
1. iloc's `_compute_lookup` suddenly returns a `slice(None)` on an empty frame, instead of a `[]`.
2. pandas backend suddenly doesn't fail on receiving `slice(None)` indexer for an empty frame.

I'm actually working on aligning these inconsistencies in #3520 and #3513, but since they aren't merged, this all looks kinda unsafe. We actually don't know how other backends would behave on an attempt to index an empty frame since there is no essential implementation for an empty core modin frame.

Maybe we should bring some safety and consistency in this PR by defaulting to pandas for both: `loc` and `iloc`? What do you think?

**Comment by naren-ponder** [comment]
> > iloc actually seems to work for the get and set without any code changes, so the only change necessary is for the loc get method.
> 
> BTW, the actual problem of why the reported bug occurs is the following way of computing a column lookup for `.loc`, for empty columns and `slice(None)` indexer it returns an empty list:
> 
> https://github.com/modin-project/modin/blob/41213581b974760928ed36f05c479937dcdfea55/modin/pandas/indexing.py#L698-L701
> 
> 
> Pandas backend is failing to process `[]` instead of a `slice(None)` since for non-full-axis grabs it tries to do some partitioning stuff under an empty frame and fails.
> It's actually pretty lucky, that the scenario from reproducer works well for the `.iloc` because the only difference from the `.loc` flow is the `_compute_lookup`, that suddenly returns a `slice(None)` indexer for an empty frame and full-axis grab, which is acceptable for pandas backend since it doesn't touch partitioning on a `slice(None)` indexer:
> 
> https://github.com/modin-project/modin/blob/41213581b974760928ed36f05c479937dcdfea55/modin/pandas/indexing.py#L817-L826
> 
> So, actually, the reason why `iloc.__getitem__` on an empty frame is working is based on two inconsistencies:
> 
> 1. iloc's `_compute_lookup` suddenly returns a `slice(None)` on an empty frame, instead of a `[]`.
> 2. pandas backend suddenly doesn't fail on receiving `slice(None)` indexer for an empty frame.
> 
> I'm actually working on aligning these inconsistencies in #3520 and #3513, but since they aren't merged, this all looks kinda unsafe. We actually don't know how other backends would behave on an attempt to index an empty frame since there is no essential implementation for an empty core modin frame.
> 
> Maybe we should bring some safety and consistency in this PR by defaulting to pandas for both: `loc` and `iloc`? What do you think?
> 
> Sorry, for probably being too picky for a small-change PR, I would also be okay with the "if it works, don't touch it" approach 😄

I was wondering about this myself, thanks for the explanation! Let me make that change -- I think it would be better stability-wise :)

File: modin/pandas/test/dataframe/test_indexing.py [hunk 2]
```diff
@@ -433,6 +433,21 @@ def test_loc_multi_index():
     df_equals(modin_df.loc[modin_df.index[:7]], pandas_df.loc[pandas_df.index[:7]])
 
 
+def test_loc_empty():
+    pandas_df1 = pandas.DataFrame(
+        np.array([i ** 2 for i in range(5)]), columns=["TIME"]
+    )
+    pandas_df2 = pandas.DataFrame(index=pandas_df1.index)
```
Comment (dchigarev) [inline modin/pandas/test/dataframe/test_indexing.py:440]: what's the point of creating a DataFrame if we use only an index from it? Am I missing something?
```suggestion
    pandas_df = pandas.DataFrame(index=range(5))
```

Comment (naren-ponder) [reply modin/pandas/test/dataframe/test_indexing.py]: You're right, not necessary for the test. I was just reproducing the exact script that was the reported bug.

**Review by dchigarev — APPROVED** [review]
LGTM, thanks for the changes!

**Review by YarShev — APPROVED** [review]
LGTM

**Pull request merged**